temporal-sdk/src/main/java/io/temporal/internal/replay/WorkflowHistoryIterator.java

```diff
@@ -77,6 +112,24 @@ public HistoryEvent next() {
       return current.next();
     }
 
+    GetWorkflowExecutionHistoryResponse response = queryWorkflowExecutionHistory(
```

Comment:
I think you can avoid this call here if line 111 calls this.hasNext

Reply:
That call is to current.hasNext(), which is just cached history.
We still should make a server call if current is empty and page token is available.
This would make it possible to safely call it.next() without calling it.hasNext() first.
That being said I've updated implementation, added tests and verified that it's working in bench, please take a look